src/ng/directive/validators.js

```diff
@@ -162,34 +167,59 @@ var requiredDirective = function() {
  *   </file>
  * </example>
  */
-var patternDirective = function() {
+var patternDirective = function($parse) {
   return {
     restrict: 'A',
     require: '?ngModel',
-    link: function(scope, elm, attr, ctrl) {
-      if (!ctrl) return;
+    compile: function(tElm, tAttr) {
+      var patternExp;
+      var parseFn;
+
+      if (tAttr.ngPattern) {
+        patternExp = tAttr.ngPattern;
 
-      var regexp, patternExp = attr.ngPattern || attr.pattern;
-      attr.$observe('pattern', function(regex) {
-        if (isString(regex) && regex.length > 0) {
-          regex = new RegExp('^' + regex + '$');
+        // ngPattern might be a scope expressions, or an inlined regex, which is not parsable.
+        // We get value of the attribute here, so we can compare the old and the new value
+        // in the observer to avoid unnecessary validations
+        try {
+          parseFn = $parse(tAttr.ngPattern);
+        } catch (e) {
+          if (/^\[\$parse:lexerr\]/.test(e.message)) {
```

Comment:
Why not re-use the logic from the `ngPattern` aliased directive?

Reply:
It wasn't handled before, because we didn't need to compare the old / new values. This is made problematic in ngPattern because it is special cased to allow RegExp (which is normally not allowed). 

You mean we should use `REGEX_STRING_REGEXP` instead?

Reply:
This still seems iffy to me :/

Reply:
@Narretz, yes, I suggest we re-use that bit of logic:

https://github.com/angular/angular.js/blob/0687a4f70a16315aacd9e74edaaac466208bb86d/src/ng/directive/attrs.js#L392-L395